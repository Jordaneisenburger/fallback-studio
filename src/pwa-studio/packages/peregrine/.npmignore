.eslintrc.js
/lib/**/__{docs,helpers,mocks,tests}__/**
/scripts
