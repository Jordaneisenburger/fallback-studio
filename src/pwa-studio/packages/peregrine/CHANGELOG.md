# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

<a name="0.5.0"></a>
# 0.5.0 (2018-08-13)


### Features

* multicasting REST API client ([#164](https://github.com/magento/pwa-studio/issues/164)) ([2852e14](https://github.com/magento/pwa-studio/commit/2852e14)), closes [#140](https://github.com/magento/pwa-studio/issues/140)
* Simulators for async testing ([#24](https://github.com/magento/pwa-studio/issues/24)) ([a380da9](https://github.com/magento/pwa-studio/commit/a380da9))
