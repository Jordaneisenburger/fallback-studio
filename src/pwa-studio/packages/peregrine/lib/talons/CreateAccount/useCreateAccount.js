import { useCallback, useMemo, useState } from 'react';
import { useMutation } from '@apollo/react-hooks';
import { useUserContext } from '@magento/peregrine/lib/context/user';

/**
 * Returns props necessary to render CreateAccount component. In particular this
 * talon handles the submission flow by first doing a pre-submisson validation
 * and then, on success, invokes the `onSubmit` prop, which is usually the action.
 *
 * @param {Object} props.initialValues initial values to sanitize and seed the form
 * @param {Function} props.onSubmit the post submit callback
 * @param {String} query the graphql query for creating the account
 * @returns {{
 *   errors: array,
 *   handleSubmit: function,
 *   isDisabled: boolean,
 *   isSignedIn: boolean,
 *   initialValues: object
 * }}
 */
export const useCreateAccount = props => {
    const { initialValues = {}, onSubmit, query } = props;
    const [isSubmitting, setIsSubmitting] = useState(false);

    const [{ isSignedIn }, { signIn }] = useUserContext();
    const [createAccount, { error }] = useMutation(query);

    const handleSubmit = useCallback(
        async formValues => {
            setIsSubmitting(true);
            try {
                // Try to create an account with the mutation.
                await createAccount({
                    variables: {
                        email: formValues.customer.email,
                        firstname: formValues.customer.firstname,
                        lastname: formValues.customer.lastname,
                        password: formValues.password
                    }
                });

                // Then sign the user in.
                await signIn({
                    username: formValues.customer.email,
                    password: formValues.password
                });

                // Finally, invoke the post-submission callback.
                onSubmit();
            } catch (error) {
                if (process.env.NODE_ENV === 'development') {
                    console.error(error);
                }
                setIsSubmitting(false);
            }
        },
        [createAccount, onSubmit, signIn]
    );

    const sanitizedInitialValues = useMemo(() => {
        const { email, firstName, lastName, ...rest } = initialValues;

        return {
            customer: { email, firstname: firstName, lastname: lastName },
            ...rest
        };
    }, [initialValues]);

    return {
        errors: (error && error.graphQLErrors) || [],
        handleSubmit,
        isDisabled: isSubmitting,
        isSignedIn,
        initialValues: sanitizedInitialValues
    };
};
