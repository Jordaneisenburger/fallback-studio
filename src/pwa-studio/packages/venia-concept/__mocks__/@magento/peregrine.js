import React from 'react';
import { BrowserPersistence } from './util/simplePersistence';

const peregrine = jest.requireActual('../../../peregrine/lib');

// re-exports
const RestApi = {
    Magento2: {
        request: jest.fn()
    }
};

const Util = { BrowserPersistence };

// hooks
const useApolloContext = jest.fn(peregrine.useApolloContext);
const useEventListener = jest.fn(peregrine.useEventListener);
const useDropdown = jest.fn(peregrine.useDropdown);
const useQuery = jest.fn(peregrine.useQuery);
const useQueryResult = jest.fn(peregrine.useQueryResult);
const useSearchParam = jest.fn(peregrine.useSearchParam);
const useWindowSize = jest.fn(peregrine.useWindowSize);
const useToasts = jest.fn(peregrine.useToasts);
// components

/**
 * the Price component from @magento/peregrine
 * has browser-specific functionality and cannot
 * currently by rendered in the test environment
 */
const Price = jest.fn().mockReturnValue(<div />);

module.exports = {
    ...peregrine,
    Price,
    RestApi,
    Util,
    useApolloContext,
    useEventListener,
    useDropdown,
    useQuery,
    useQueryResult,
    useSearchParam,
    useWindowSize,

    useToasts
};
