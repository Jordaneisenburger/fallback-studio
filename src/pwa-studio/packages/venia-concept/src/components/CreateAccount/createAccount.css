.root {
    display: grid;
    gap: 0.5rem;
    justify-items: stretch;
    padding: 0 1.5rem;
}

.actions {
    display: grid;
    justify-items: center;
    margin-top: 1rem;
    padding: 1rem 0;
}

.error {
    color: rgb(var(--venia-error));
    text-align: center;
}

.lead {
    margin: 1rem 0;
    text-align: center;
}

.subscribe {
    margin-left: -0.5rem;
}
