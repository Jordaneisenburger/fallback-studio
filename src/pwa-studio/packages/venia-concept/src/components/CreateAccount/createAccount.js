import React, { Component } from 'react';
import { func, shape, string } from 'prop-types';
import { Form } from 'informed';

import classify from 'src/classify';
import Button from 'src/components/Button';
import Checkbox from 'src/components/Checkbox';
import Field from 'src/components/Field';
import TextInput from 'src/components/TextInput';

import combine from 'src/util/combineValidators';
import {
    validateEmail,
    isRequired,
    validatePassword,
    validateConfirmPassword,
    hasLengthAtLeast
} from 'src/util/formValidators';

import defaultClasses from './createAccount.css';

class CreateAccount extends Component {
    static propTypes = {
        classes: shape({
            actions: string,
            error: string,
            lead: string,
            root: string,
            subscribe: string
        }),
        createAccountError: shape({
            message: string
        }),
        initialValues: shape({
            email: string,
            firstName: string,
            lastName: string
        }),
        onSubmit: func
    };

    static defaultProps = {
        initialValues: {}
    };

    get errorMessage() {
        const { createAccountError } = this.props;

        if (createAccountError) {
            const errorIsEmpty = Object.keys(createAccountError).length === 0;
            if (!errorIsEmpty) {
                return 'An error occurred. Please try again.';
            }
        }
    }

    get initialValues() {
        const { initialValues } = this.props;
        const { email, firstName, lastName, ...rest } = initialValues;

        return {
            customer: { email, firstname: firstName, lastname: lastName },
            ...rest
        };
    }

    handleSubmit = values => {
        const { onSubmit } = this.props;

        if (typeof onSubmit === 'function') {
            onSubmit(values);
        }
    };

    render() {
        const { errorMessage, handleSubmit, initialValues, props } = this;
        const { classes } = props;

        return (
            <Form
                className={classes.root}
                initialValues={initialValues}
                onSubmit={handleSubmit}
            >
                <h3 className={classes.lead}>
                    {`Check out faster, use multiple addresses, track
                         orders and more by creating an account!`}
                </h3>
                <Field label="First Name" required={true}>
                    <TextInput
                        field="customer.firstname"
                        autoComplete="given-name"
                        validate={isRequired}
                        validateOnBlur
                    />
                </Field>
                <Field label="Last Name" required={true}>
                    <TextInput
                        field="customer.lastname"
                        autoComplete="family-name"
                        validate={isRequired}
                        validateOnBlur
                    />
                </Field>
                <Field label="Email" required={true}>
                    <TextInput
                        field="customer.email"
                        autoComplete="email"
                        validate={combine([isRequired, validateEmail])}
                        validateOnBlur
                    />
                </Field>
                <Field label="Password" required={true}>
                    <TextInput
                        field="password"
                        type="password"
                        autoComplete="new-password"
                        validate={combine([
                            isRequired,
                            [hasLengthAtLeast, 8],
                            validatePassword
                        ])}
                        validateOnBlur
                    />
                </Field>
                <Field label="Confirm Password" required={true}>
                    <TextInput
                        field="confirm"
                        type="password"
                        validate={combine([
                            isRequired,
                            validateConfirmPassword
                        ])}
                        validateOnBlur
                    />
                </Field>
                <div className={classes.subscribe}>
                    <Checkbox
                        field="subscribe"
                        label="Subscribe to news and updates"
                    />
                </div>
                <div className={classes.error}>{errorMessage}</div>
                <div className={classes.actions}>
                    <Button type="submit" priority="high">
                        {'Submit'}
                    </Button>
                </div>
            </Form>
        );
    }
}

export default classify(defaultClasses)(CreateAccount);
