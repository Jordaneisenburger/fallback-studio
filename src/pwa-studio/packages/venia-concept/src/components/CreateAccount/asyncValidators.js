import { RestApi } from '@magento/peregrine';

const { request } = RestApi.Magento2;

export const validateEmail = async value => {
    try {
        const body = {
            customerEmail: value,
            website_id: null
        };

        // response is a boolean
        const available = await request('/rest/V1/customers/isEmailAvailable', {
            method: 'POST',
            body: JSON.stringify(body)
        });

        return !available ? 'This email address is not available.' : null;
    } catch (error) {
        throw 'An error occurred while looking up this email address.';
    }
};
