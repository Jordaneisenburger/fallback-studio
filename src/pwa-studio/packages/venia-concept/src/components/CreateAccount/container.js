import { connect } from '@magento/venia-drivers';
import CreateAccount from './createAccount';

const mapStateToProps = ({ user }) => {
    const { createAccountError } = user;
    return {
        createAccountError
    };
};

export default connect(mapStateToProps)(CreateAccount);
