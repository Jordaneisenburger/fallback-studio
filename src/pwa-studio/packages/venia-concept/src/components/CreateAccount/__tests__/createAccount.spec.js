import React from 'react';
import { act } from 'react-test-renderer';
import waitForExpect from 'wait-for-expect';
import { Form } from 'informed';
import { createTestInstance } from '@magento/peregrine';

import CreateAccount from '../createAccount';

jest.mock('src/util/formValidators');

export const submitCallback = jest.fn();

test('renders the correct tree', () => {
    const tree = createTestInstance(<CreateAccount />).toJSON();

    expect(tree).toMatchSnapshot();
});

test('has a submit handler', () => {
    const { root } = createTestInstance(<CreateAccount />);

    const { instance } = root.children[0];

    expect(instance.handleSubmit).toBeInstanceOf(Function);
});

test('attaches the submit handler', () => {
    const { root } = createTestInstance(<CreateAccount />);

    const { instance } = root.children[0];
    const { onSubmit } = root.findByType(Form).props;

    expect(onSubmit).toBe(instance.handleSubmit);
});

test('calls onSubmit if validation passes', async () => {
    const { root } = createTestInstance(
        <CreateAccount onSubmit={submitCallback} />
    );

    const form = root.findByType(Form);
    const { formApi } = form.instance;

    // touch fields, call validators, call onSubmit
    act(() => {
        formApi.submitForm();
    });

    // await async validation
    await waitForExpect(() => {
        const { errors } = form.instance.controller.state;

        expect(Object.keys(errors)).toHaveLength(0);
        expect(submitCallback).toHaveBeenCalledTimes(1);
    });
});
