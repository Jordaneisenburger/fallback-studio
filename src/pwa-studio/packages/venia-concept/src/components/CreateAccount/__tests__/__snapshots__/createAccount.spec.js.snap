// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the correct tree 1`] = `
<form
  onReset={[Function]}
  onSubmit={[Function]}
>
  <h3>
    Check out faster, use multiple addresses, track
                         orders and more by creating an account!
  </h3>
  <div>
    <span>
      <span />
      First Name
    </span>
    <span
      style={
        Object {
          "--iconsAfter": 0,
          "--iconsBefore": 0,
        }
      }
    >
      <span>
        <input
          autoComplete="given-name"
          name="customer.firstname"
          onBlur={[Function]}
          onChange={[Function]}
          value=""
        />
      </span>
      <span />
      <span />
    </span>
    <p />
  </div>
  <div>
    <span>
      <span />
      Last Name
    </span>
    <span
      style={
        Object {
          "--iconsAfter": 0,
          "--iconsBefore": 0,
        }
      }
    >
      <span>
        <input
          autoComplete="family-name"
          name="customer.lastname"
          onBlur={[Function]}
          onChange={[Function]}
          value=""
        />
      </span>
      <span />
      <span />
    </span>
    <p />
  </div>
  <div>
    <span>
      <span />
      Email
    </span>
    <span
      style={
        Object {
          "--iconsAfter": 0,
          "--iconsBefore": 0,
        }
      }
    >
      <span>
        <input
          autoComplete="email"
          name="customer.email"
          onBlur={[Function]}
          onChange={[Function]}
          value=""
        />
      </span>
      <span />
      <span />
    </span>
    <p />
  </div>
  <div>
    <span>
      <span />
      Password
    </span>
    <span
      style={
        Object {
          "--iconsAfter": 0,
          "--iconsBefore": 0,
        }
      }
    >
      <span>
        <input
          autoComplete="new-password"
          name="password"
          onBlur={[Function]}
          onChange={[Function]}
          type="password"
          value=""
        />
      </span>
      <span />
      <span />
    </span>
    <p />
  </div>
  <div>
    <span>
      <span />
      Confirm Password
    </span>
    <span
      style={
        Object {
          "--iconsAfter": 0,
          "--iconsBefore": 0,
        }
      }
    >
      <span>
        <input
          name="confirm"
          onBlur={[Function]}
          onChange={[Function]}
          type="password"
          value=""
        />
      </span>
      <span />
      <span />
    </span>
    <p />
  </div>
  <div>
    <label>
      <span />
      <input
        checked={false}
        name="subscribe"
        onBlur={[Function]}
        onChange={[Function]}
        type="checkbox"
      />
      <span>
        Subscribe to news and updates
      </span>
    </label>
    <p />
  </div>
  <div />
  <div>
    <button
      type="submit"
    >
      <span>
        Submit
      </span>
    </button>
  </div>
</form>
`;
